File: sdks/php/src/Template.php
<?php
namespace Autonoma\Sdk;

use DateTime;
use DateTimeZone;

class Template
{
    private const TEMPLATE_RE = '/\{\{(.+?)\}\}/';

    /** Resolve all {{...}} expressions in a value. Handles strings, arrays, dicts recursively. */
    public static function resolveTemplate(mixed $value, array $ctx): mixed
    {
        if (is_string($value)) {
            return self::resolveString($value, $ctx);
        }
        if (is_array($value)) {
            if (self::isAssoc($value)) {
                $result = [];
                foreach ($value as $k => $v) {
                    $result[$k] = self::resolveTemplate($v, $ctx);
                }
                return $result;
            }
            return array_map(fn($v) => self::resolveTemplate($v, $ctx), $value);
        }
        return $value;
    }

    private static function resolveString(string $s, array $ctx): mixed
    {
        // If the entire string is a single expression, return raw value (preserving type)
        if (preg_match('/^\{\{(.+?)\}\}$/', $s, $m)) {
            return self::evaluateExpression(trim($m[1]), $ctx);
        }
        // Otherwise, interpolate expressions into the string
        return preg_replace_callback(self::TEMPLATE_RE, function ($match) use ($ctx) {
            return (string) self::evaluateExpression(trim($match[1]), $ctx);
        }, $s);
    }

    private static function evaluateExpression(string $expr, array $ctx): mixed
    {
        if ($expr === 'testRunId') {
            return $ctx['testRunId'] ?? $ctx['test_run_id'] ?? '';
        }
        if ($expr === 'index') {
            return $ctx['index'] ?? 0;
        }
        if ($expr === 'index1') {
            return ($ctx['index'] ?? 0) + 1;
        }
        if ($expr === 'now()') {
            $dt = new DateTime('now', new DateTimeZone('UTC'));
            return $dt->format('Y-m-d\TH:i:s.v\Z');
        }

        // cycle([...])
        if (preg_match('/^cycle\(\[(.+)\]\)$/', $expr, $m)) {
            $items = self::parseArrayLiteral($m[1]);
            $index = $ctx['index'] ?? 0;
            return $items[$index % count($items)];
        }

        // pick([...])
        if (preg_match('/^pick\(\[(.+)\]\)$/', $expr, $m)) {
            $items = self::parseArrayLiteral($m[1]);
            return $items[array_rand($items)];
        }

        // random.int(a,b)
        if (preg_match('/^random\.int\((\d+),\s*(\d+)\)$/', $expr, $m)) {
            return random_int((int) $m[1], (int) $m[2]);
        }

        // random.float(a,b)
        if (preg_match('/^random\.float\((\d+(?:\.\d+)?),\s*(\d+(?:\.\d+)?)\)$/', $expr, $m)) {
            $min = (float) $m[1];
            $max = (float) $m[2];
            return $min + (mt_rand() / mt_getrandmax()) * ($max - $min);
        }

        // daysAgo(n)
        if (preg_match('/^daysAgo\((\d+)\)$/', $expr, $m)) {
            $n = (int) $m[1];
            $dt = new DateTime('now', new DateTimeZone('UTC'));
            $dt->modify("-{$n} days");
            return $dt->format('Y-m-d\TH:i:s.v\Z');
        }

        throw new \RuntimeException("Template error: unknown expression '{$expr}'");
    }

    /** @return string[] */
    private static function parseArrayLiteral(string $raw): array
    {
        $items = [];
        foreach (explode(',', $raw) as $s) {
            $s = trim($s);
            if ((str_starts_with($s, "'") && str_ends_with($s, "'")) ||
                (str_starts_with($s, '"') && str_ends_with($s, '"'))) {
                $s = substr($s, 1, -1);
            }
            $items[] = $s;
        }
        return $items;
    }

    private static function isAssoc(array $arr): bool
    {
        if (empty($arr)) return false;
        return array_keys($arr) !== range(0, count($arr) - 1);
    }
}

File: sdks/rust/src/template.rs
//! Template expression resolution for {{...}} expressions in entity specs.

use chrono::{Duration, Utc};
use rand::Rng;
use regex::Regex;
use serde_json::Value;
use std::sync::LazyLock;

static TEMPLATE_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\{\{(.+?)\}\}").unwrap());
static FULL_TEMPLATE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\{\{(.+?)\}\}$").unwrap());
static CYCLE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^cycle\(\[(.+)\]\)$").unwrap());
static PICK_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^pick\(\[(.+)\]\)$").unwrap());
static RAND_INT_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^random\.int\((\d+),\s*(\d+)\)$").unwrap());
static RAND_FLOAT_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^random\.float\((\d+(?:\.\d+)?),\s*(\d+(?:\.\d+)?)\)$").unwrap());
static DAYS_AGO_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^daysAgo\((\d+)\)$").unwrap());

/// Resolve all {{...}} expressions in a value. Handles strings, dicts, lists recursively.
pub fn resolve_template(value: &Value, ctx: &Value) -> Value {
    match value {
        Value::String(s) => resolve_string(s, ctx),
        Value::Array(arr) => Value::Array(arr.iter().map(|v| resolve_template(v, ctx)).collect()),
        Value::Object(map) => {
            let resolved: serde_json::Map<String, Value> = map
                .iter()
                .map(|(k, v)| (k.clone(), resolve_template(v, ctx)))
                .collect();
            Value::Object(resolved)
        }
        other => other.clone(),
    }
}

fn resolve_string(s: &str, ctx: &Value) -> Value {
    // If the entire string is a single expression, return raw value (preserving type)
    if let Some(caps) = FULL_TEMPLATE_RE.captures(s) {
        let expr = caps.get(1).unwrap().as_str().trim();
        return evaluate_expression(expr, ctx);
    }

    // Otherwise, interpolate expressions into the string
    let result = TEMPLATE_RE.replace_all(s, |caps: &regex::Captures| {
        let expr = caps.get(1).unwrap().as_str().trim();
        let val = evaluate_expression(expr, ctx);
        match val {
            Value::String(s) => s,
            Value::Number(n) => n.to_string(),
            Value::Bool(b) => b.to_string(),
            Value::Null => "null".to_string(),
            other => other.to_string(),
        }
    });
    Value::String(result.into_owned())
}

fn evaluate_expression(expr: &str, ctx: &Value) -> Value {
    if expr == "testRunId" {
        let val = ctx
            .get("testRunId")
            .or_else(|| ctx.get("test_run_id"))
            .cloned()
            .unwrap_or(Value::String(String::new()));
        return val;
    }

    if expr == "index" {
        let idx = ctx
            .get("index")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);
        return Value::Number(idx.into());
    }

    if expr == "index1" {
        let idx = ctx
            .get("index")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);
        return Value::Number((idx + 1).into());
    }

    if expr == "now()" {
        let now = Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string();
        return Value::String(now);
    }

    // cycle([...])
    if let Some(caps) = CYCLE_RE.captures(expr) {
        let items = parse_array_literal(caps.get(1).unwrap().as_str());
        let index = ctx
            .get("index")
            .and_then(|v| v.as_i64())
            .unwrap_or(0) as usize;
        let picked = &items[index % items.len()];
        return Value::String(picked.clone());
    }

    // pick([...])
    if let Some(caps) = PICK_RE.captures(expr) {
        let items = parse_array_literal(caps.get(1).unwrap().as_str());
        let mut rng = rand::thread_rng();
        let idx = rng.gen_range(0..items.len());
        return Value::String(items[idx].clone());
    }

    // random.int(a,b)
    if let Some(caps) = RAND_INT_RE.captures(expr) {
        let min: i64 = caps.get(1).unwrap().as_str().parse().unwrap();
        let max: i64 = caps.get(2).unwrap().as_str().parse().unwrap();
        let mut rng = rand::thread_rng();
        let val = rng.gen_range(min..=max);
        return Value::Number(val.into());
    }

    // random.float(a,b)
    if let Some(caps) = RAND_FLOAT_RE.captures(expr) {
        let min: f64 = caps.get(1).unwrap().as_str().parse().unwrap();
        let max: f64 = caps.get(2).unwrap().as_str().parse().unwrap();
        let mut rng = rand::thread_rng();
        let val = rng.gen_range(min..max);
        return serde_json::Number::from_f64(val)
            .map(Value::Number)
            .unwrap_or(Value::Null);
    }

    // daysAgo(n)
    if let Some(caps) = DAYS_AGO_RE.captures(expr) {
        let n: i64 = caps.get(1).unwrap().as_str().parse().unwrap();
        let dt = Utc::now() - Duration::days(n);
        let iso = dt.format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string();
        return Value::String(iso);
    }

    Value::String(format!("Template error: unknown expression '{}'", expr))
}

fn parse_array_literal(raw: &str) -> Vec<String> {
    raw.split(',')
        .map(|s| {
            let s = s.trim();
            if (s.starts_with('\'') && s.ends_with('\''))
                || (s.starts_with('"') && s.ends_with('"'))
            {
                s[1..s.len() - 1].to_string()
            } else {
                s.to_string()
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn resolves_test_run_id() {
        let result = resolve_template(
            &json!("{{testRunId}}"),
            &json!({"testRunId": "run-123"}),
        );
        assert_eq!(result, json!("run-123"));
    }

    #[test]
    fn resolves_index() {
        let result = resolve_template(&json!("{{index}}"), &json!({"index": 2}));
        assert_eq!(result, json!(2));
    }

    #[test]
    fn resolves_cycle() {
        let result = resolve_template(
            &json!("{{cycle(['a','b','c'])}}"),
            &json!({"index": 4}),
        );
        assert_eq!(result, json!("b"));
    }

    #[test]
    fn preserves_non_string_types() {
        let result = resolve_template(&json!(42), &json!({}));
        assert_eq!(result, json!(42));
    }

    #[test]
    fn string_interpolation() {
        let result = resolve_template(
            &json!("hello-{{testRunId}}-world"),
            &json!({"testRunId": "abc"}),
        );
        assert_eq!(result, json!("hello-abc-world"));
    }
}
